File: src/edictum_server/schemas/bundles.py
"""Request and response schemas for bundle and deployment endpoints."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class BundleUploadRequest(BaseModel):
    """Upload a new contract bundle (raw YAML string)."""

    yaml_content: str


class BundleResponse(BaseModel):
    """Serialized bundle returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    version: int
    revision_hash: str
    signature_hex: str | None = None
    source_hub_slug: str | None = None
    source_hub_revision: str | None = None
    uploaded_by: str
    created_at: datetime


class BundleWithDeploymentsResponse(BundleResponse):
    """Bundle response enriched with deployed environment names."""

    deployed_envs: list[str] = []


class DeployRequest(BaseModel):
    """Request body for deploying a bundle to an environment."""

    env: str


class DeploymentResponse(BaseModel):
    """Serialized deployment returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    env: str
    bundle_version: int
    deployed_by: str
    created_at: datetime

File: dashboard/src/lib/api/bundles.ts
import { API_BASE, ApiError, request } from "./client"

export interface BundleResponse {
  id: string
  tenant_id: string
  version: number
  revision_hash: string
  signature_hex: string | null
  uploaded_by: string
  created_at: string
}

export interface BundleWithDeployments extends BundleResponse {
  deployed_envs: string[]
}

export interface DeploymentResponse {
  id: string
  env: string
  bundle_version: number
  deployed_by: string
  created_at: string
}

export function listBundles() {
  return request<BundleWithDeployments[]>("/bundles")
}

export function uploadBundle(yamlContent: string) {
  return request<BundleResponse>("/bundles", {
    method: "POST",
    body: JSON.stringify({ yaml_content: yamlContent }),
  })
}

export function deployBundle(version: number, env: string) {
  return request<DeploymentResponse>(`/bundles/${version}/deploy`, {
    method: "POST",
    body: JSON.stringify({ env }),
  })
}

export async function getBundleYaml(version: number): Promise<string> {
  const res = await fetch(`${API_BASE}/bundles/${version}/yaml`, {
    credentials: "include",
  })
  if (!res.ok) {
    const body = await res.text()
    throw new ApiError(res.status, body)
  }
  return res.text()
}

export function getCurrentBundle(env: string) {
  return request<BundleResponse>(`/bundles/current?env=${encodeURIComponent(env)}`)
}
